// angular-input-counter/esm5/lib/counter.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CounterComponent } from './counter.component';
import { MatInputModule } from '@angular/material/input';
import { MatIconModule } from '@angular/material/icon';
import { MatButtonModule } from '@angular/material/button';
import { OnlyDigitsDirective } from './counter.directive';
import { FormsModule } from '@angular/forms';
import * as ɵngcc0 from '@angular/core';
var AngularCounterModule = /** @class */ (function () {
    function AngularCounterModule() {
    }
AngularCounterModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: AngularCounterModule });
AngularCounterModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function AngularCounterModule_Factory(t) { return new (t || AngularCounterModule)(); }, imports: [[
            MatInputModule,
            MatIconModule,
            MatButtonModule, FormsModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(AngularCounterModule, { declarations: function () { return [CounterComponent, OnlyDigitsDirective]; }, imports: function () { return [MatInputModule,
        MatIconModule,
        MatButtonModule, FormsModule]; }, exports: function () { return [CounterComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AngularCounterModule, [{
        type: NgModule,
        args: [{
                declarations: [CounterComponent, OnlyDigitsDirective],
                imports: [
                    MatInputModule,
                    MatIconModule,
                    MatButtonModule, FormsModule
                ],
                exports: [CounterComponent]
            }]
    }], function () { return []; }, null); })();
    return AngularCounterModule;
}());
export { AngularCounterModule };
