// angular-input-counter/esm5/lib/counter.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef } from '@angular/core';
//only-digits
import * as ɵngcc0 from '@angular/core';
var OnlyDigitsDirective = /** @class */ (function () {
    function OnlyDigitsDirective(el) {
        this.el = el;
        this.el.nativeElement.onkeypress = (/**
         * @param {?} evt
         * @return {?}
         */
        function (evt) {
            if (evt.which < 48 || evt.which > 57) {
                evt.preventDefault();
            }
        });
    }
    /** @nocollapse */
    OnlyDigitsDirective.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
OnlyDigitsDirective.ɵfac = function OnlyDigitsDirective_Factory(t) { return new (t || OnlyDigitsDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef)); };
OnlyDigitsDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: OnlyDigitsDirective, selectors: [["", "only-digits", ""]] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(OnlyDigitsDirective, [{
        type: Directive,
        args: [{
                selector: '[only-digits]'
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }]; }, null); })();
    return OnlyDigitsDirective;
}());
export { OnlyDigitsDirective };
if (false) {
    /** @type {?} */
    OnlyDigitsDirective.prototype.el;
}
